Initial OpenACC port of mpas_atm_update_bdy_tend

abishekg7 opened the pull request:
Initial OpenACC port of `mpas_atm_update_bdy_tend`. 

This port is required to keep `state` and `tend` variables from LBCs eventually resident on GPUs.

gdicker1 reviewed (commented):
The changes make sense and GPU runs of this PR and a reference before these changes match.

I'll hold off on full approval for any changes Michael requests or for any changes to the commit history that may come.

mgduda commented on src/core_atmosphere/dynamics/mpas_atm_boundaries.F:
```diff
@@ -176,26 +180,86 @@ subroutine mpas_atm_update_bdy_tend(clock, streamManager, block, firstCall, ierr
+        !$acc end kernels
+
+        !$acc parallel
+        ! Compute lbc_rho_zz
```
We could either remove this comment, or remove the same comment on line 217. I'd suggest just removing this line.

abishekg7 replied:
Removed redundant comment on this line

mgduda commented on src/core_atmosphere/dynamics/mpas_atm_boundaries.F:
```diff
@@ -225,15 +289,58 @@ subroutine mpas_atm_update_bdy_tend(clock, streamManager, block, firstCall, ierr
+            !$acc loop gang vector collapse(2)
+             do iEdge=1,nEdges+1
+                 do k=1,nVertLevels
+                     lbc_tend_rho_edge(k,iEdge) = (rho_edge(k,iEdge) - lbc_tend_rho_edge(k,iEdge)) * dt
```
Was there a specific reason for computing `lbc_tend_u` and `lbc_tend_ru` in the same loop, but using a separate loop for `lbc_tend_rho_edge`?

abishekg7 replied:
I wasn't sure if splitting up the loops would result in better performance on GPUs, but I'll just keep it simple and consolidate all similar loops for now and investigate optimizations later.

mgduda commented on src/core_atmosphere/dynamics/mpas_atm_boundaries.F:
```diff
@@ -225,15 +289,58 @@ subroutine mpas_atm_update_bdy_tend(clock, streamManager, block, firstCall, ierr
+            do iCell=1,nCells+1
+                do k=1,nVertLevels
+                    lbc_tend_rho_zz(k,iCell) = (rho_zz(k,iCell) - lbc_tend_rho_zz(k,iCell)) * dt
+                    lbc_tend_rho(k,iCell) = (rho(k,iCell) - lbc_tend_rho(k,iCell)) * dt
```
Similar to the comment above, would there be an advantage, or is there a disadvantage, to combining the computation of `lbc_tend_rho_zz` and `lbc_tend_rho` in the same loops where we compute `lbc_tend_theta` and `lbc_tend_rtheta_m`?

abishekg7 replied:
Consolidating loops as mentioned above.

mgduda commented on src/core_atmosphere/dynamics/mpas_atm_boundaries.F:
```diff
@@ -225,15 +289,58 @@ subroutine mpas_atm_update_bdy_tend(clock, streamManager, block, firstCall, ierr
+                !$acc loop vector collapse(2)
+                do k=1,nVertLevels
+                    do j = 1,nScalars
+                    lbc_tend_scalars(j,k,iCell) = (scalars(j,k,iCell) - lbc_tend_scalars(j,k,iCell)) * dt
```
Let's indent the body of this loop.

abishekg7 replied:
Done.

mgduda commented on src/core_atmosphere/dynamics/mpas_atm_boundaries.F:
```diff
@@ -225,15 +289,58 @@ subroutine mpas_atm_update_bdy_tend(clock, streamManager, block, firstCall, ierr
+            end do
+            !$acc end parallel
+
+            !$acc parallel default(present)
```
Could this `parallel` region be combined with the `parallel` region above it?

abishekg7 replied:
Done.

mgduda commented on src/core_atmosphere/dynamics/mpas_atm_boundaries.F:
```diff
@@ -176,26 +180,86 @@ subroutine mpas_atm_update_bdy_tend(clock, streamManager, block, firstCall, ierr
+            call mpas_pool_get_array(lbc, 'lbc_rtheta_m', lbc_tend_rtheta_m, 1)
+            call mpas_pool_get_array(lbc, 'lbc_rho_zz', lbc_tend_rho_zz, 1)
+            call mpas_pool_get_array(lbc, 'lbc_rho', lbc_tend_rho, 1)
+            call mpas_pool_get_array(lbc, 'lbc_scalars', lbc_tend_scalars, 1)
```
Since we now get pointers to arrays here, I think it would be fine to include in this PR the removal of redundant calls to `mpas_pool_get_array` below (beginning around line 270).

mgduda commented on src/core_atmosphere/dynamics/mpas_atm_boundaries.F:
```diff
@@ -167,6 +170,7 @@ subroutine mpas_atm_update_bdy_tend(clock, streamManager, block, firstCall, ierr
         ! Compute any derived fields from those that were read from the lbc_in stream
         !
         call mpas_pool_get_array(lbc, 'lbc_u', u, 2)
+        call mpas_pool_get_array(lbc, 'lbc_w', w, 2)
```
It's a minor detail, but can we move this below the call to `mpas_pool_get_array` for `lbc_rho_edge` so that the order of variables is consistent in this block and in the block beginning at line 192?

abishekg7 replied:
Done.

abishekg7 commented:
Thanks for the review! I'll address the comments, but just wanted to note that this PR may not be a priority for merging to `develop` with host-device memory transfers consolidated around every call to dynamics. But it will be needed eventually.

abishekg7 commented:
Addressed the review comments. I did have to rebase with the latest `develop` to get bit identical results (with the latest `develop`). Also noticing the same for #1315 that I'm working on. Checking if commits from earlier this week are causing some of these differences.

mgduda commented on src/core_atmosphere/dynamics/mpas_atm_boundaries.F:
```diff
@@ -249,6 +321,17 @@ subroutine mpas_atm_update_bdy_tend(clock, streamManager, block, firstCall, ierr
 
+        MPAS_ACC_TIMER_START('mpas_atm_update_bdy_tend [ACC_data_xfer]')
+        if (.not. firstCall) then
+           !$acc exit data copyout(lbc_tend_u, lbc_tend_ru, lbc_tend_rho_edge, lbc_tend_w, &
```
Could we add one more space of indentation here to match the indentation in the two lines that follow?

abishekg7 replied:
Done. thanks for the catch!

mgduda commented on src/core_atmosphere/dynamics/mpas_atm_boundaries.F:
```diff
@@ -169,71 +172,140 @@ subroutine mpas_atm_update_bdy_tend(clock, streamManager, block, firstCall, ierr
-        rho_zz(:,:) = rho(:,:) / zz(:,:)
+        !$acc end kernels
+
+        !$acc parallel
```
Should we add `default(present)` to this `parallel` directive (and also to others below)?

abishekg7 replied:
Thanks! Done and I've also converted the `kernels` statement to a `parallel` region, as it is not a temporary loop. Combining these `parallel` regions is a little tricky, so I've left them separate as it was.

mgduda reviewed (approved):
Looks good, thanks! Once you've had a chance to rework the commit history, I'll go ahead and merge this.

abishekg7 commented:
> Looks good, thanks! Once you've had a chance to rework the commit history, I'll go ahead and merge this.

Done. And thanks for the review!

gdicker1 reviewed (approved):
Looks good!
